File: general-interview-prep/javascript/sufficientSubset.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} limit
 * @return {TreeNode}
 */

var sufficientSubset = function(root, limit) {
    if (root === null) {
        return null;
    }
    if (root.left === null && root.right === null) {
        return root.val < limit ? null : root;
    }

    root.left = sufficientSubset(root.left, limit - root.val);
    root.right = sufficientSubset(root.right, limit - root.val);
    return root.left === root.right ? null : root;
};

File: general-interview-prep/golang/sufficientSubset.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func sufficientSubset(root *TreeNode, limit int) *TreeNode {
	if root == nil {
			return nil
	}
	
	if root.Left == nil && root.Right == nil {
			if root.Val < limit {
					return nil
			} else {
					return root
			}
	}
	
	root.Left = sufficientSubset(root.Left, limit - root.Val)
	root.Right = sufficientSubset(root.Right, limit - root.Val)
	
	if root.Left == root.Right {
			return nil
	}
	
	return root
}